netlogo-gui/src/main/app/Tabs.scala

```diff
@@ -93,132 +105,154 @@ class Tabs(val workspace:    GUIWorkspace,
+  def getSource(filename: String): String = getTabWithFilename(Right(filename)).map(_.innerSource).orNull
 
-  private def getTabWithFilename(name: String): Option[TemporaryCodeTab] =
-    // start at 3 because 0, 1, and 2 are the permanent tabs
-    (3 until getTabCount)
-      .map(getComponentAt)
-      .collect{case tab: TemporaryCodeTab => tab}
-      .find(_.filename == name)
+  def getTabWithFilename(filename: Filename): Option[TemporaryCodeTab] =
+    externalFileTabs find (_.filename == filename)
 
-  def newTemporaryFile() { addNewTab(TemporaryCodeTab.NewFile, false) }
+  def newExternalFile() = addNewTab(Left(I18N.gui.getN("tabs.external.new", TemporaryCodeTab.serialNum: Integer)))
 
-  def addNewTab(name: String, fileMustExist: Boolean) {
-    val tab = new TemporaryCodeTab(workspace, this, name, fileMustExist, codeTab.smartTabbingEnabled)
-    addTab(stripPath(name), tab)
-    addMenuItem(getTabCount() - 1, stripPath(name))
-    org.nlogo.window.Event.rehash()
-    tab.includesMenu.updateVisibility()
+  def openExternalFile(filename: String) =
+    getTabWithFilename(Right(filename)) match {
+      case Some(tab) => setSelectedComponent(tab)
+      case _ => addNewTab(Right(filename))
+    }
+
+  def addNewTab(name: Filename) = {
+    val tab = new TemporaryCodeTab(workspace, this, name, codeTab.smartTabbingEnabled)
+    externalFileTabs += tab
+    addTab(tab.filenameForDisplay, tab)
+    addMenuItem(getTabCount - 1, tab.filenameForDisplay)
+    Event.rehash()
     setSelectedComponent(tab)
     // if I just call requestFocus the tab never gets the focus request because it's not yet
     // visible.  There might be a more swing appropriate way to do this but I can't figure it out
     // (if you know it feel free to fix) ev 7/24/07
-    org.nlogo.awt.EventQueue.invokeLater( () => requestFocus() )
-  }
-
-  def saveExternalFiles() {
-    (3 until getTabCount)
-      .map(getComponentAt)
-      .collect{case tab: TemporaryCodeTab => tab}
-      .foreach(_.doSave())
+    EventQueue.invokeLater( () => requestFocus() )
   }
 
-  def saveTemporaryFile(filename: String) {
-    getTabWithFilename(filename) foreach { tab =>
-      val index = getIndexOfComponent(tab)
-      setTitleAt(index, stripPath(filename))
-      tabActions(index).putValue(Action.NAME, filename)
-    }
-  }
+  def saveExternalFiles() = externalFileTabs foreach (_.doSave)
 
   def getIndexOfComponent(tab: CodeTab): Int =
     (0 until getTabCount).find(n => getComponentAt(n) == tab).get
 
-  def closeTemporaryFile(filename: String) {
+  def closeExternalFile(filename: Filename): Unit =
     getTabWithFilename(filename) foreach { tab =>
       val index = getIndexOfComponent(tab)
       remove(tab)
       removeMenuItem(index)
+      externalFileTabs -= tab
     }
-  }
 
   def forAllCodeTabs(fn: CodeTab => Unit) =
-    getComponents collect { case tab: CodeTab => tab } foreach (fn)
+    (externalFileTabs.asInstanceOf[Set[CodeTab]] + codeTab) foreach fn
```

Comment:
Yes, otherwise you cannot add `codeTab` (of type `MainCodeTab`) to a `Set[TemporaryCodeTab]`, so I convert it to a `Set[CodeTab]`